### PR TITLE
put out appropriate message when collections are deleted by batch.

### LINKED ISSUE #1788: Collections: Batch delete should show message if fails
### Descriptive summary

If any selected collection fails to delete, there should be a message stating that some of the collections failed to delete due to lack of permissions.

Users can select any collection they can see in Dashboard -> Collections.  They may not have edit access to all those collections.  If they select collections with read access and click `Delete collections`, only collections with edit access will be deleted.  The read only access collections will silently continue to exist. 

### Rationale

Users should receive feedback when an action they take is not fully completed.

### Expected behavior

When attempting to delete a read only collection or a non-empty admin set, a message should be displayed stating the attempt failed.

> "One or more of the selected collections could not be deleted. You cannot delete a collection for which you have read-only access. In addition, you cannot delete a collection of the type "Administrative set" unless that collection is empty."

Note: Sites can customize the name of admin sets, so get the exact name from the collection type with `machine_id='admin_set'`.

Required changes:
- [ ] show checkbox for all collection types, including admin sets
- [ ] during delete processing, ensure user can delete
  - [ ] for admin-set, check that user has edit_access and the admin set is empty
  - [ ] for all other types, check that user has edit_access
  - [ ] delete all that can be deleted
- [ ] if there were any that could not be deleted, show a dialog with the message text as shown above

### Actual behavior

When attempting to delete a read only collection, no message is displaying stating the attempt failed.

### Steps to reproduce the behavior

1. Create two users: user 1 (an admin user), user 2 (a regular user)
1. Login as user 1, the admin
1. Create a collection with title 'Managed Collection' and give user 2 'Manage' access to the collection
1. Create a collection with title 'Viewable Collection' and give user 2 'View' access to the collection
1. Login as user 2
1. Dashboard -> Collections -> Managed Collections

- [ ] Verify that 'Managed Collection' and 'Viewable Collection' are both in the list of collections.

1. select all -> Delete collections

- [ ] Verify that 'Managed Collection' is no longer in the list.
- [ ] Verify that 'Viewable Collection' is in the list.
- [ ] Verify that a message stating that not all selected collections could be deleted is displayed as a flash message.  (Exact text in section 'Expected behavior' above.)

1. Login as user 1, the admin
1. Create an empty admin set with title 'Empty Admin Set'
1. Create another admin set with title 'Admin Set with Works'
1. Create a work and put it in 'Admin Set with Works'
1. Dashboard -> Collections -> Your Collections

- [ ] Verify that 'Empty Admin Set' and 'Admin Set with Works' are both in the list of collections.

1. select all -> Delete collections

- [ ] Verify that 'Empty Admin Set' is no longer in the list.
- [ ] Verify that 'Admin Set with Works' is in the list.
- [ ] Verify that a message stating that not all selected collections could be deleted is displayed as a flash message.  (Exact text in section 'Expected behavior' above.)

### Related work

#1510 - PR #1716 - batch delete collections
#1708 UI/UX question regarding checkboxes for AdminSets on Collections index page

### PR DESCRIPTION
**Pull request by blancoj**
Fixes #1788

### CODE REVIEW

File: app/controllers/hyrax/batch_edits_controller.rb [hunk 1]
```diff
@@ -39,7 +39,15 @@ def destroy_collection
         obj = ActiveFedora::Base.find(doc_id, cast: true)
         obj.destroy
       end
-      flash[:notice] = "Batch delete complete"
+
+      admin_set_name = "Administrative Set"
```
Comment (jcoyne) [inline app/controllers/hyrax/batch_edits_controller.rb]: Please use i18n.

File: app/controllers/hyrax/batch_edits_controller.rb [hunk 2]
```diff
@@ -39,7 +39,15 @@ def destroy_collection
...
+      admin_set_name = AdminSet.first.title.first unless AdminSet.first.nil?
```
Comment (jcoyne) [inline app/controllers/hyrax/batch_edits_controller.rb]: Of all the admin sets, why are you choosing the first one? Why are you testing for nil?

File: app/controllers/hyrax/batch_edits_controller.rb [hunk 3]
```diff
@@ -39,7 +39,15 @@ def destroy_collection
...
+
+      flash[:notice] = if flash[:notice].nil?
+                         "Batch delete complete"
```
Comment (jcoyne) [inline app/controllers/hyrax/batch_edits_controller.rb]: Please use i18n.

File: app/controllers/hyrax/batch_edits_controller.rb [hunk 4]
```diff
@@ -39,7 +39,15 @@ def destroy_collection
...
+                       else
+                         I18n.t('errors.messages.batch_delete_collections_error', admin_set_name: admin_set_name)
```
Comment (jcoyne) [inline app/controllers/hyrax/batch_edits_controller.rb]: Won't this show up even when the delete is successful?  What about when the batch is empty? Will it overwrite the 'Select something first' message?

File: app/controllers/concerns/hyrax/collections/accepts_batches.rb [hunk 5]
```diff
@@ -36,7 +36,7 @@ def filter_docs_with_edit_access!
 
         def filter_docs_with_access!(access_type = :edit)
           no_permissions = []
-          if batch.empty?
+          if batch.empty? && flash[:notice].nil?
```
Comment (jcoyne) [inline app/controllers/concerns/hyrax/collections/accepts_batches.rb]: How would `flash[:notice]` get set prior to running this code?

File: config/locales/hyrax.en.yml [hunk 6]
```diff
@@ -31,6 +31,7 @@ en:
       carrierwave_processing_error: "Cannot resize image."
       extension_blacklist_error: "You are not allowed to upload %{extension} files, prohibited types: %{prohibited_types}"
       extension_whitelist_error: "You are not allowed to upload %{extension} files, allowed types: %{allowed_types}"
+      batch_delete_collections_error: "One or more of the selected collections could not be deleted. You cannot delete a collection for which you have read-only access. In addition, you cannot delete a collection of the type \"%{admin_set_name}\" unless that collection is empty."
```
Comment (jcoyne) [inline config/locales/hyrax.en.yml]: This key doesn't give much of a clue to where this message is used.  How about `hyrax.batch_edit.delete_collection.type_error`?

File: app/controllers/hyrax/batch_edits_controller.rb (hunk 1 above)
Comment (blancoj) [reply app/controllers/hyrax/batch_edits_controller.rb]: yes!

File: app/controllers/hyrax/batch_edits_controller.rb (hunk 2 above)
Comment (blancoj) [reply app/controllers/hyrax/batch_edits_controller.rb]: Changing it to: `admin_set_type_name = Hyrax::CollectionType.find_or_create_admin_set_type.title`

File: app/controllers/hyrax/batch_edits_controller.rb (hunk 3 above)
Comment (blancoj) [reply app/controllers/hyrax/batch_edits_controller.rb]: yes

File: config/locales/hyrax.en.yml (hunk 6 above)
Comment (blancoj) [reply config/locales/hyrax.en.yml]: yes!

File: spec/controllers/hyrax/batch_edits_controller_spec.rb [hunk 7]
```diff
@@ -127,6 +129,7 @@
         expect { Collection.find(collection1.id) }.not_to raise_error(Ldp::Gone)
         expect { Collection.find(collection2.id) }.not_to raise_error(Ldp::Gone)
         expect { Collection.find(collection3.id) }.to raise_error(Ldp::Gone)
+        expect(flash[:notice]).to eq("One or more of the selected collections could not be deleted. You cannot delete a collection for which you have read-only access. In addition, you cannot delete a collection of the type \"Admin Set\" unless that collection is empty.")
       end
     end
   end
```
Comment (elrayle) [inline spec/controllers/hyrax/batch_edits_controller_spec.rb]: The tests look good.  I like the coverage of collections.  Can you add one that tests including an empty admin set and an admin set with works?

File: spec/controllers/hyrax/batch_edits_controller_spec.rb [hunk 8]
```diff
@@ -93,11 +97,26 @@
     let(:curation_concern) { create(:work1, user: user) }
 
     context 'when user has edit access' do
-      it "deletes collections with and without works in it" do
-        controller.batch = [collection1.id, collection2.id]
+      it "deletes collections with and without works in it and an empty admin set." do
+        controller.batch = [collection1.id, collection2.id, empty_adminset.id]
+        delete :destroy_collection, params: { update_type: "delete_all" }
+        expect { Collection.find(collection1.id) }.to raise_error(Ldp::Gone)
```
Comment (jcoyne) [inline spec/controllers/hyrax/batch_edits_controller_spec.rb]: can you check `expect(collection1).not_to exist(collection1.id)` rather than trapping for errors?

File: app/controllers/hyrax/batch_edits_controller.rb [hunk 9]
```diff
@@ -35,11 +35,19 @@ def check_for_empty!
     end
 
     def destroy_collection
+      # This flag is especially useful to catch non empty admin sets.
+      success = true
       batch.each do |doc_id|
         obj = ActiveFedora::Base.find(doc_id, cast: true)
-        obj.destroy
+        success = false if obj.destroy == false
```
Comment (jcoyne) [inline app/controllers/hyrax/batch_edits_controller.rb]: Why are we checking for `obj.destroy == false`?  What could cause this situation?  Do you have a test for this change?

File: app/controllers/hyrax/batch_edits_controller.rb [hunk 10]
```diff
@@ -35,11 +35,19 @@ def check_for_empty!
...
       end
-      flash[:notice] = "Batch delete complete"
+
+      admin_set_type_name = Hyrax::CollectionType.find_or_create_admin_set_type.title
+      flash[:notice] = if flash[:notice].nil? && success
```
Comment (jcoyne) [inline app/controllers/hyrax/batch_edits_controller.rb]: Why are we checking to make sure `flash[:notice].nil?`?

File: app/controllers/hyrax/batch_edits_controller.rb [hunk 11]
```diff
@@ -35,11 +35,19 @@ def check_for_empty!
...
+                         I18n.t('hyrax.batch_edit.delete_complete.success')
+                       else
+                         I18n.t('hyrax.batch_edit.delete_complete.type_error', admin_set_name: admin_set_type_name)
```
Comment (jcoyne) [inline app/controllers/hyrax/batch_edits_controller.rb]: This will be returned if `flash[:notice]` is set, regardless of the value of `success`. Is that your intention?

File: app/views/hyrax/dashboard/collections/_list_collections.html.erb [hunk 12]
```diff
@@ -1,6 +1,6 @@
 <% id = collection_presenter.id %>
 <tr id="document_<%= id %>">
-  <td><% unless collection_presenter.solr_document.admin_set? %><input type="checkbox" name="batch_document_ids[]" id="batch_document_<%= id %>" value="<%= id %>" class="batch_document_selector" /><% end %></td>
+  <td><input type="checkbox" name="batch_document_ids[]" id="batch_document_<%= id %>" value="<%= id %>" class="batch_document_selector" /></td>
```
Comment (jcoyne) [inline app/views/hyrax/dashboard/collections/_list_collections.html.erb]: I don't understand what this change has to do with changing the message that is displayed when deleting a batch. Can you explain?

File: spec/controllers/hyrax/batch_edits_controller_spec.rb [hunk 13]
```diff
@@ -74,6 +74,10 @@
   describe "#destroy_collection" do
     let(:user) { create(:user) }
 
+    let!(:empty_adminset) { create(:admin_set, title: ['Empty Admin Set'], edit_users: [user], creator: [user.user_key], with_permission_template: true) }
+    let!(:adminset) { create(:admin_set, title: ['Admin Set with Work'], edit_users: [user], creator: [user.user_key], with_permission_template: true) }
```
Comment (jcoyne) [inline spec/controllers/hyrax/batch_edits_controller_spec.rb]: The `let!` isn't required here `let` would be fine because `adminset` is referenced on line 79

File: spec/controllers/hyrax/batch_edits_controller_spec.rb [hunk 14]
```diff
@@ -93,11 +97,26 @@
...
+      end
+    end
+
+    context 'when user has edit access' do
+      it "deletes collections with and without works in it and a none empty admin set is not deleted." do
+        controller.batch = [collection1.id, collection2.id, adminset.id]
         delete :destroy_collection, params: { update_type: "delete_all" }
         expect { Collection.find(collection1.id) }.to raise_error(Ldp::Gone)
         expect { Collection.find(collection2.id) }.to raise_error(Ldp::Gone)
+        expect { AdminSet.find(adminset.id) }.not_to raise_error(Ldp::Gone)
```
Comment (jcoyne) [inline spec/controllers/hyrax/batch_edits_controller_spec.rb]: prefer `expect(AdminSet).not_to exist(adminset.id)`

File: spec/controllers/hyrax/batch_edits_controller_spec.rb [hunk 15]
```diff
@@ -114,19 +133,23 @@
         allow(controller).to receive(:current_user).and_return(user2)
       end
 
-      it "fails to delete collections when user does not have edit access" do
-        controller.batch = [collection1.id, collection3.id]
+      it "fails to delete collections, and admin sets when user does not have edit access" do
```
Comment (jcoyne) [inline spec/controllers/hyrax/batch_edits_controller_spec.rb]: Should this be separate test cases?  With one test I'm not sure we can be certain both cases are covered.

File: spec/controllers/hyrax/batch_edits_controller_spec.rb [hunk 16]
```diff
@@ -114,19 +133,23 @@
...
         expect { Collection.find(collection1.id) }.not_to raise_error(Ldp::Gone)
         expect { Collection.find(collection2.id) }.not_to raise_error(Ldp::Gone)
+        expect { AdminSet.find(adminset.id) }.not_to raise_error(Ldp::Gone)
+        expect { AdminSet.find(empty_adminset.id) }.not_to raise_error(Ldp::Gone)
+        expect(flash[:notice]).to include("One or more of the selected collections could not be deleted")
       end
 
       it "deletes collections where user has edit access, failing to delete those where user does not have edit access" do
-        controller.batch = [collection1.id, collection3.id]
+        controller.batch = [collection1.id, collection2.id, collection3.id]
```
Comment (jcoyne) [inline spec/controllers/hyrax/batch_edits_controller_spec.rb]: why was it important to add another collection to the batch?

File: app/controllers/hyrax/batch_edits_controller.rb [hunk 17]
```diff
@@ -6,7 +6,7 @@ class BatchEditsController < ApplicationController
 
     before_action :build_breadcrumbs, only: :edit
     before_action :filter_docs_with_access!, only: [:edit, :update, :destroy_collection]
-    before_action :check_for_empty!, only: [:edit, :update, :destroy_collection]
+    before_action :check_for_empty!, only: [:edit, :update]
```
Comment (jcoyne) [inline app/controllers/hyrax/batch_edits_controller.rb]: Why are we not checking to make sure they selected something?

**Comment by elrayle** [comment]
This was replaced by PR #2870.  Closing.
